adapter/src/utils/useLocale.js

```diff
@@ -3,16 +3,45 @@ import i18n from '@dhis2/d2-i18n'
 import moment from 'moment'
 import { useState, useEffect } from 'react'
 
-i18n.setDefaultNamespace('default')
+const I18N_NAMESPACE = 'default'
+i18n.setDefaultNamespace(I18N_NAMESPACE)
 
-const simplifyLocale = (locale) => {
-    const idx = locale.indexOf('-')
-    if (idx === -1) {
+const transformJavaLocale = (locale) => {
+    return locale.replace('_', '-')
+}
+
+// if translation resources aren't found for the given locale, try shorter
+// versions of the locale
+// e.g. 'pt_BR_Cyrl_asdf' => 'pt_BR', or 'ar-NotFound' => 'ar'
+const validateLocaleByBundle = (locale) => {
+    if (i18n.hasResourceBundle(locale, I18N_NAMESPACE)) {
+        return locale
+    }
+
+    console.log(`Translations for locale ${locale} not found`)
+
+    // see if we can try basic versions of the locale
+    // (e.g. 'ar' instead of 'ar_IQ')
+    const match = /[_-]/.exec(locale)
+    if (!match) {
         return locale
     }
-    return locale.substr(0, idx)
+
+    const separator = match[0] // '-' or '_'
+    const splitLocale = locale.split(separator)
+    for (let i = splitLocale.length - 1; i > 0; i--) {
+        const shorterLocale = splitLocale.slice(0, i).join(separator)
+        if (i18n.hasResourceBundle(shorterLocale, I18N_NAMESPACE)) {
+            return shorterLocale
+        }
+        console.log(`Translations for locale ${shorterLocale} not found`)
+    }
+
```

Comment:
for this logic for getting the shorter version, instead of doing it manually by splitting .. couldn't we use the Locale methods, i.e.

```
const locale = new Intl.Locale('ja-Jpan-JP');

console.log(locale.language, locale.region, locale.script);
// will give us language = ja, region = JP, script = Jpan
// and we can try to get the locale for language or `${language}-${region}``
```

Reply:
Hmm yeah that's a nice helper function -- if that works, I think that can help make that code look a lot more semantic 👍 Implementing it is a bit complex given the existing translation files we have to work with, for example `uz_UZ_Cyrl` isn't in the order that `Intl.Locale` wants, but I'll see what I can do (when using hyphens, `new Intl.Locale('uz-UZ-Cyrl')` throws an error)

Here's a screenshot from the Dashboard app's locales:

![Screenshot 2024-01-19 at 1 42 58 PM](https://github.com/dhis2/app-platform/assets/49666798/5f3e8c10-f388-446e-b1be-9f1c25cf573f)